lib/reporting/cloudwatch_client.rb

```diff
@@ -0,0 +1,219 @@
+require 'aws-sdk-cloudwatchlogs'
+require 'concurrent-ruby'
+require 'reporting/cloudwatch_query'
+require 'ruby-progressbar'
+
+module Reporting
+  class CloudwatchClient
+    DEFAULT_NUM_THREADS = 5
+    DEFAULT_WAIT_DURATION = 3
+
+    attr_reader :num_threads, :wait_duration, :slice_interval, :logger
+
+    # @param [Boolean] ensure_complete_logs when true, will detect when queries return exactly
+    #  10,000 rows (Cloudwatch Insights max limit) and then recursively split the query window into
+    #  two queries until we're certain we've queried all rows
+    # @param [ActiveSupport::Duration,Boolean,nil] slice_interval starting interval to split up and
+    #  query across, or something falsy to indicate not to slice the query
+    # @param [Boolean,nil,IO,#fileno] progress whether or not to show progress, and which IO
+    #  to send send the progress bar to (defaults to STDERR)
+    def initialize(
+      ensure_complete_logs: true,
+      num_threads: DEFAULT_NUM_THREADS,
+      wait_duration: DEFAULT_WAIT_DURATION,
```

Comment:
Speeding up tests is the main one, but also if we ever get consistent timing for the CW queries, we could tune the sleep to wait on the more precisely instead of wasting time